### PR TITLE
BUG: Raise TypeError for float->timedelta promotion

### LINKED ISSUE #16566: BUG: timedelta and float promotion sometimes works
The following examples do not lead to symmetric results:
```
np.promote_types("float32", "m8")
# returns "m8", which is probably wrong
np.promote_types("m8", "float32")
# raises TypeError.
```
The same is true for all other floating point types.  I wonder if there is some strange logic to this, as a hack to allow certain ufunc calls, but I did not find one yet.

### PR DESCRIPTION
**Pull request by anirudh2290**
Fixes https://github.com/numpy/numpy/issues/16566

### CODE REVIEW

**Comment by seberg** [comment]
@jorisvandenbossche in my branch on array-coercion, the one failure that is obviously related to the change (I expect most others are just due to the master branch/other issues), is a test like this:
```
    def test_infer_dtype_timedelta_with_na(self, na_value, delta):
        # starts with nan
        arr = np.array([na_value, delta])
        assert lib.infer_dtype(arr, skipna=True) == "timedelta"

        arr = np.array([na_value, delta, na_value])
        assert lib.infer_dtype(arr, skipna=True) == "timedelta"
```
Roughly, this test fails on my branch, because it currently promotes the other way around `[delta, na_value]` (so I can swap that, and it should be the same).  This PR will deprecate the whole thing, which here would mean that in the future you will get an `object` array.  I assume pandas would be happy about the general fix?

But, it also means you would get deprecation/future warnings in this type of array coercion...  so the more general question is, whether we try to hack a more userfriendly warning (I did something similar in my start on string+numeric promotion), or whether we rather just rip off the band-aid here, considering that it is inconsistent and rather a bug to begin with.

I am starting to favor that approach.  With regard to array-coercion, it will just make things work (by returning object arrays), which does not seem terrible. And since it was an error previously, it seems unlikely that someone `try:/excepted:` it.

**Comment by jbrockmendel** [comment]
what are na_value and delta in the failing test(s)?

**Comment by seberg** [comment]
Ah sorry, its `np.nan` and an arbitrary timedelta (nothing special).  The problem is that sometimes `np.promote_types(float, timedelta64)` gives `timedelta64`, and sometimes it gives an error (which for array-coercion means `object` and `object` is what pandas needs I guess)

**Comment by jbrockmendel** [comment]
Can you post a traceback?  its hard to guess where lib.infer_dtype would go wrong here

**Comment by seberg** [comment]
Oh, my bad again, but I guess that also proofs that pandas is just not dependend on NumPy here. The line that changes – that I am worried about – is the `np.array()` line...

**Comment by seberg** [comment]
Sorry to be clear:
```
In [2]: np.array([np.nan, np.timedelta64(1, "D")])                                                                      
Out[2]: array([nan, numpy.timedelta64(1,'D')], dtype=object)

In [3]: np.array([np.timedelta64(1, "D"), np.nan])                                                                      ValueError: Could not convert object to NumPy timedelta
```
Is the current behaviour. This PR makes to always return `object`. If we were to deprecate that though, we might end up with a warning in the second case. Which seems undesirable?

**Comment by jbrockmendel** [comment]
> This PR makes to always return object

This seems reasonable to me.  It isnt obvious how that would break anything within pandas

**Comment by seberg** [comment]
@anirudh2290 I think I would like to try and put this in.  Maybe we should add a short release note just in case?

**Comment by anirudh2290** [comment]
@seberg done !

File: doc/release/upcoming_changes/16592.compatibility.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+float->timedelta promotion will raise a TypeError
+-------------------------------------------------
+``np.promote_types("float32", "m8")`` and other float to timedelta promotion
+which used to return "m8" (timedelta) before, will now raise a TypeError.
```
Comment (seberg) [inline doc/release/upcoming_changes/16592.compatibility.rst:4]: I am wondering if we mention that it aligns it with `np.promote_types("m8", "float32")` to ease peoples mind that it really should not be a large change.  E.g. could also smuggle the word "consistently" in there ;). "Float and timedelta promotion consistently raises TypeError"?

But, this is also fine, so maybe I will just merge, or update myself later.

Comment (anirudh2290) [reply doc/release/upcoming_changes/16592.compatibility.rst]: thanks ! i have made a change to address this.

Comment (seberg) [reply doc/release/upcoming_changes/16592.compatibility.rst]: Thanks Anirudh, lets put this in.

**Comment by seberg** [comment]
Some strange test failure. I think they must be due to these changes being based on a not quite up-to-date branch. In any case nothing to worry about/unrelated to the change.

**Comment by seberg** [comment]
Ehhh. My bad, I completely missed that this is also the (bad?) behaviour for all or some integers:
```
In [1]: np.promote_types("m8", "uint64") 
TypeError: invalid type promotion

In [2]: np.promote_types("uint64", "m8")  
Out[2]: dtype('<m8')
```
And I guess we can again simply change it without any real downstream impact, since ufuncs don't use promotion often to begin with, and even if, have the loops defined correctly for dtypes...

**Comment by anirudh2290** [comment]
> Ehhh. My bad, I completely missed that this is also the (bad?) behaviour for all or some integers:
> 
> ```
> In [1]: np.promote_types("m8", "uint64") 
> TypeError: invalid type promotion
> 
> In [2]: np.promote_types("uint64", "m8")  
> Out[2]: dtype('<m8')
> ```
> 
> And I guess we can again simply change it without any real downstream impact, since ufuncs don't use promotion often to begin with, and even if, have the loops defined correctly for dtypes...

aah, sure. let me open another pr for this !

**Pull request merged**